packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,420 @@
+            merkleDistributor.setWindow(
+                reward.maximumRewardAmount,
+                address(reward.rewardToken),
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+            emit RewardDistributed(
+                reward.sponsor,
+                reward.rewardToken,
+                proposal.rewardIndex,
+                proposalIndex,
+                reward.maximumRewardAmount,
+                proposal.merkleRoot,
+                proposal.ipfsHash
+            );
+        } else emit ProposalRejected(proposal.rewardIndex, proposalIndex);
+
+        // Delete resolved proposal from storage. This also avoids double-spend for approved proposals.
+        delete proposals[proposalIndex];
+        emit ProposalDeleted(proposal.rewardIndex, proposalIndex);
+    }
+
+    /**
+     * @notice Allows any caller to delete distribution that was rejected by the Optimistic Oracle.
+     */
+    function deleteRejectedDistribution(uint256 proposalIndex) external nonReentrant() {}
+
+    /********************************************
+     *          MAINTENANCE FUNCTIONS           *
+     ********************************************/
+
+    /**
+     * @notice Sets address of MerkleDistributor contract that will be used for rewards distribution.
+     * MerkleDistributor address can only be set once.
+     * @dev It is expected that the deployer first deploys MekleDistributor contract and transfers its ownership to
+     * the OptimisticDistributor contract and then calls `setMerkleDistributor` on the OptimisticDistributor pointing
+     * on now owned MekleDistributor contract.
+     * @param _merkleDistributor Address of the owned MerkleDistributor contract.
+     */
+    function setMerkleDistributor(MerkleDistributor _merkleDistributor) external nonReentrant() {
+        require(address(merkleDistributor) == address(0), "MerkleDistributor already set");
+        require(_merkleDistributor.owner() == address(this), "MerkleDistributor not owned");
+
+        merkleDistributor = _merkleDistributor;
+        emit MerkleDistributorSet(address(_merkleDistributor));
+    }
+
+    /**
+     * @notice Updates the address stored in this contract for the OptimisticOracle and the Store to the latest
+     * versions set in the Finder. Also pull finalFee from Store contract.
+     * @dev There is no risk of leaving this function public for anyone to call as in all cases we want the addresses
+     * in this contract to map to the latest version in the Finder and store the latest final fee.
+     */
+    function syncUmaEcosystemParams() public nonReentrant() {
+        store = _getStore();
+        finalFee = store.computeFinalFee(address(bondToken)).rawValue;
+        optimisticOracle = _getOptimisticOracle();
+    }
+
+    // Can be overriden for testing.
+    function getCurrentTime() public view virtual returns (uint256) {
```

Comment:
Alternative would be to inherit Testable in production code (like in LSP), but currently I used the approach as in optimistic-rewarder. Which approach shall we keep going forward?

Reply:
both are pretty similar. IMHO doing the `testable` version is more consistant and cleaner as it means you won't need to have the `test` contract for `OptimisticDistributor` as you can simply deploy a `Timer` in your tests and then pass it in.